## COSC 1210
### print(), input(), value storage and chaos

Objectives
- use the chaos function to illustrate the importance of value storage

In [ ]:

def main():
    a = 1/7
    b = a + a + a + a + a + a + a
    print(b)
main()

0.9999999999999998


##### Explain why the function above provides an answer that is different than the algebraic solution
bacause the value of a is saved as a decimal number, there will be a rounding error

In [2]:
def main():
    a = 1/7
    b = a * 7
    c = a + a + a + a + a + a + a
    print("b is", b)
    print("c is", c)
    print(b - c)
main()

b is 1.0
c is 0.9999999999999998
2.220446049250313e-16


##### Explain why b and c provide different answers
b is only a single operation
c contains six operations, each operation in c can accumulate rounding errors


##### How large is the difference between c and b?
0.0000000000000002



### rounding floats
Mathematically, what will be the value of b?
When calculated by python, what is the value of b?

In [8]:
def main():
    print("This program illustrates a chaotic function")
    iterations = 10
    x1 = 0.15
    for i in range(iterations):
        x1 = 3.9 * x1 * (1 - x1)
    print("For", iterations, "iterations, x1 is:", x1)
main()

This program illustrates a chaotic function
For 10 iterations, x1 is: 0.9735732406265619


##### If you calculate x1 for 10 iterations vs 100 iterations, what answers do you get?

for 10 iterations: 0.9735732406265619

for 100 iterations: 0.5126527786207501

##### Why are these different?
both goes through different number of iterations


In [11]:
def main():
    print("This program illustrates a chaotic function")
    iterations = 100
    x1 = 0.15
    x2 = 0.15
    for i in range(iterations):
        x1 = 3.9 * x1 * (1 - x1)
        x2 = 3.9 * (x2 - x2 * x2)
    print("For", iterations, "iterations, x1 is now", x1, "and x2 is now", x2)
main()

This program illustrates a chaotic function
For 100 iterations, x1 is now 0.5126527786207501 and x2 is now 0.16411537941360743


##### If you calculate x1 and x2 for 10 iterations vs 100 iterations, what answers do you get?
x1 10 iterations: 0.9735732406265619
x2 10 iterations: 0.9735732406265659

x1 100 iterations: 0.5126527786207501
x2 100 iterations: 0.16411537941360743

Why are these different?
x1 has 2 variables while x2 has 3 variables, the number of variable in both are different, and the more operatios it goes through the more rounding errors accumulate, thus there is a difference between the answers

In [9]:
### Rewrite the chaos function so that the use can enter x
def main():
    print("This program illustrates a chaotic function")
    iterations = 100
    user_answer = input("enter a number")
    user_answer = eval(user_answer)
    x1 = user_answer
    x2 = user_answer
    for i in range(iterations):
        x1 = 3.9 * x1 * (1 - x1)
        x2 = 3.9 * (x2 - x2 * x2)
    print("For", iterations, "iterations, x1 is now", x1, "and x2 is now", x2)
main()

This program illustrates a chaotic function
For 100 iterations, x1 is now -inf and x2 is now -inf
